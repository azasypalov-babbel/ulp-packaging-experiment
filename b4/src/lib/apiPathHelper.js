export const buildPath = ({ version, locale, uuid, learnLanguageAlpha3 }, rest = '') => {
  if (!version) {
    throw 'Missing required param "version"';
  }
  if (!locale) {
    throw 'Missing required param "locale"';
  }
  if (!uuid) {
    throw 'Missing required param "uuid"';
  }
  if (!learnLanguageAlpha3) {
    throw 'Missing required param "learnLanguageAlpha3"';
  }

  return `v${version}/${locale}/accounts/${uuid}/learn_languages/${learnLanguageAlpha3}/${rest}`;
};

export const buildPathV5 = ({ version, locale, uuid, contentReleaseId, learnLanguageAlpha3 }, rest = '') => {
  if (!version) {
    throw 'Missing required param "version"';
  }
  if (!locale) {
    throw 'Missing required param "locale"';
  }
  if (!uuid) {
    throw 'Missing required param "uuid"';
  }
  if (!learnLanguageAlpha3) {
    throw 'Missing required param "learnLanguageAlpha3"';
  }
  if (!contentReleaseId) {
    throw 'Missing required param "contentReleaseId"';
  }

  return `v${version}/${locale}/accounts/${uuid}/content_releases/${contentReleaseId}` +
    `/learn_languages/${learnLanguageAlpha3}/${rest}`;
};
