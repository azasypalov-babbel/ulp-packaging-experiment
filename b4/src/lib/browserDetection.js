// please note,
// that IE11 now returns undefined again for window.chrome
// and new Opera 30 outputs true for window.chrome
// but needs to check if window.opr is not undefined
// and new IE Edge outputs to true now for window.chrome
// and if not iOS Chrome check
// so use the below updated condition

// based on https://stackoverflow.com/a/13348618
const isChromium = window.chrome;
const agent = window.navigator.userAgent.toLowerCase();
const vendorName = window.navigator.vendor;
const isOpera = typeof window.opr !== 'undefined';
const isIEedge = agent.indexOf('edge') > -1 || agent.indexOf('edg') > -1;
let isChrome = false;

if (
  isChromium !== null &&
  typeof isChromium !== 'undefined' &&
  vendorName === 'Google Inc.' &&
  isOpera === false &&
  isIEedge === false
) {
  isChrome = true;
}

export { isChrome };
