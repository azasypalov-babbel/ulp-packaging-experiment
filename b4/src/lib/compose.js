export const compose = (...functionsToCompose) => (argumentsOfFunctions) => {
  return functionsToCompose.reduceRight((result, functionToCompose) => {
    return functionToCompose(result);
  }, argumentsOfFunctions);
};
