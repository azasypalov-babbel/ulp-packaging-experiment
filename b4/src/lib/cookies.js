function identity(x) {
  return x;
}

function convertFeatureStrToObject(elem) {
  const elements = elem.trim().split('=');
  return {
    [elements[0]]: elements[1]
  };
}

function arrayToObjectReducer(acc, element) {
  return {
    ...acc,
    ...element
  };
}

function getCookies() {
  return document?.cookie.split(';')
    .filter(identity)
    .map(convertFeatureStrToObject)
    .reduce(arrayToObjectReducer, {});
}

function getCookie(cookie) {
  const cookies = getCookies();

  return cookies && cookies[cookie];
}

function setCookie(key, value, expiryDate) {
  document.cookie = `${key}=${value};expires=${expiryDate.toUTCString()};path=/`;
}

export default {
  setCookie,
  getCookies,
  getCookie
};
