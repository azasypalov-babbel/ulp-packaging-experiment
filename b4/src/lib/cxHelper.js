import baseCX from 'classnames';

const PREFIX = 'loy';

const cx = (...args) => {
  const classNames = baseCX(args);

  return classNames.split(' ')
    .map((className) => `${PREFIX}-${className}`)
    .join(' ');
};

export default cx;
