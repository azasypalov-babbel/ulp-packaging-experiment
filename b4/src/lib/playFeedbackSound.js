import services from '../services';
import correctFeedbackSound from '@assets/sounds/babbel_correct.mp3';
import incorrectFeedbackSound from '@assets/sounds/babbel_wrong.mp3';
import endscreenFeedbackSound from '@assets/sounds/endscreen.mp3';


const { soundService, settingsService } = services;

soundService.preload(correctFeedbackSound);
soundService.preload(incorrectFeedbackSound);

export const FEEDBACK_SOUND = {
  CORRECT: 'CORRECT',
  INCORRECT: 'INCORRECT',
  END: 'END',
  forItemSolved: (solved) => {
    return solved
      ? FEEDBACK_SOUND.CORRECT
      : FEEDBACK_SOUND.INCORRECT;
  }
};

const feedbackSounds = {
  [FEEDBACK_SOUND.CORRECT]: correctFeedbackSound,
  [FEEDBACK_SOUND.INCORRECT]: incorrectFeedbackSound,
  [FEEDBACK_SOUND.END]: endscreenFeedbackSound
};

/**
 * If isFeedbackSoundEnabled, it will call the soundService.play method
 * If not, it will call onEnded (which is otherwise called by soundService.play)
 *
 * @param {str} feedbackSound} - Must be one of FEEDBACK_SOUND.CORRECT / INCORRECT / END.
 * @param {Object} options - Options
 * @param {function} options.onPlay - Callback when playback starts
 * @param {function} options.onEnded - Callback when playback finished
 */
const playFeedbackSound = (feedbackSound, options = {}) => {
  const {
    onEnded = () => {}
  } = options;

  if (settingsService.isFeedbackSoundEnabled) {
    const instance = soundService.getInstance(feedbackSounds[feedbackSound]);
    const unsub = instance.subscribe(() => {
      if (!soundService.isPlaying(instance)) {
        onEnded();
        unsub();
      }
    });

    instance.play();
  } else {
    onEnded();
  }
};

export default playFeedbackSound;
