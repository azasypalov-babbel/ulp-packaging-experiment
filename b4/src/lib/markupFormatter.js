import { compose } from '../lib/compose';
import {
  formatParentheses,
  getFirstCorrectSolution,
  removeGapFormatting,
  removeFirstParanthesesAndGapsFormatting,
  markupStringToPlainText
} from '@lessonnine/babbel-markup-helper.js';

const commonFormat = compose(
  removeGapFormatting,
  formatParentheses,
  markupStringToPlainText
);

export const formatDL = (s) => {
  if (!s) return '';

  return compose(
    removeFirstParanthesesAndGapsFormatting,
    commonFormat
  )(s);
};

export const formatLL = (s) => {
  if (!s) return '';

  return compose(
    commonFormat,
    getFirstCorrectSolution
  )(s);
};

export const stripParantheses = (text) => text.replace(/(\([\S]+\))/g, '').trim();

