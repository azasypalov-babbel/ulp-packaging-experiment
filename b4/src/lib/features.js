import cookies from './cookies';
const { setCookie, getCookies } = cookies;

let featureStore = {};

const defaultAllowedFeatures = [
  'is_demo',
  'is_redux_logger',
  'is_redux_devtools',
  'is_unlocked',
  'is_cat_preview',
  'is_analyzer_lib_recognizer',
  'is_b3_vocabulary_trainer',
  'is_web_speech_mock',
  'is_native_speech',
  'is_verbose_logging',
  'is_transliteration',
  'is_fake_caret',
  'is_selection_helper',
  'is_puzzle_helper',
  'is_debug',
  'is_refer_a_friend',
  'is_zendesk'
];

const defaultNonPersistentFeatures = [
  'is_demo',
  'is_analyzer_lib_recognizer',
  'is_b3_vocabulary_trainer',
  'is_fake_caret',
  'is_selection_helper',
  'is_puzzle_helper',
  'is_zendesk'
];

function filterFeatures(features, allowedFeatures) {
  const filteredFeatures = {};
  Object.keys(features).forEach(function(key) {
    if (allowedFeatures.includes(key) && (features[key] === 'on' || features[key] === 'off')) {
      filteredFeatures[key] = features[key];
    }
  });
  return filteredFeatures;
}

function filterPersistentFeatures(features, nonPersistentFeatures) {
  const filteredFeatures = {};
  Object.keys(features).forEach(function(key) {
    if (!nonPersistentFeatures.includes(key)) {
      filteredFeatures[key] = features[key];
    }
  });
  return filteredFeatures;
}

export function configure(
  features,
  allowedFeatures = defaultAllowedFeatures,
  nonPersistentFeatures = defaultNonPersistentFeatures
) {
  featureStore = filterFeatures(features, allowedFeatures);

  const persistantFeatures = filterPersistentFeatures(
    filterFeatures(getCookies(), allowedFeatures),
    nonPersistentFeatures
  );

  Object.entries(persistantFeatures)
    .forEach(function([key, value]) {
      if (!(key in featureStore)) {
        featureStore[key] = value;
      }
    });

  const expiryDate = new Date();
  const oneYearMs = 365 * 24 * 60 * 60 * 1000;
  expiryDate.setTime(expiryDate.getTime() + oneYearMs);

  Object.entries(filterPersistentFeatures(featureStore, nonPersistentFeatures))
    .forEach(function([key, value]) {
      setCookie(key, value, expiryDate);
    });
}

export function get(featureName) {
  if (featureStore[featureName] === 'on') {
    return true;
  } else if (featureStore[featureName] === 'off') {
    return false;
  } else {
    return null;
  }
}

export function getActiveFeatures() {
  return Object.keys(featureStore)
    .filter((key) => featureStore[key] === 'on');
}

export function isWebview() {
  return process.env.WEBVIEW === '1';
}

export function isAndroid() {
  return (
    isWebview()
    && typeof window.androidListener?.postMessage === 'function'
  );
}

export function isiOS() {
  return (
    isWebview() &&
    typeof window.webkit?.messageHandlers?.iosListener?.postMessage === 'function'
  );
}
