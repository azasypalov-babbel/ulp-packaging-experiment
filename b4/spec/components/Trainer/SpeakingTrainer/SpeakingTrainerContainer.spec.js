import React from 'react';
import {
  SpeakingTrainerContainer,
  mapStateToProps
} from '../../../../src/components/Trainer/SpeakingTrainer/SpeakingTrainerContainer';
import { shallow } from 'enzyme';
import {
  formatParentheses,
  getFirstCorrectSolution,
  removeGapFormatting,
  markupStringToPlainText,
  stripCurlyBracesAndContent
} from '@lessonnine/babbel-markup-helper.js';

jest.mock('../../../../src/lib/localStorage', () => ({
  getLocalStorage: () => ({
    set: jest.fn(),
    get: jest.fn()
  })
}));
jest.mock('../../../../src/components/Trainer/SpeakingTrainer/SpeakingTrainer');

import { formatDL } from '../../../../src/lib/markupFormatter';

jest.mock('../../../../src/lib/markupFormatter', () => ({
  formatDL: jest.fn((arg) => arg)
}));

jest.mock('@lessonnine/babbel-markup-helper.js', () => ({
  formatParentheses: jest.fn((s) => s),
  getFirstCorrectSolution: jest.fn((s) => s),
  removeGapFormatting: jest.fn((s) => s),
  markupStringToPlainText: jest.fn((s) => s),
  stripCurlyBracesAndContent: jest.fn((s) => s)
}));

import playFeedbackSound from '../../../../src/lib/playFeedbackSound';
jest.mock('../../../../src/lib/playFeedbackSound');

const mockAudio = {
  playSound: jest.fn((soundId, { onEnded }) => {
    if (onEnded) onEnded();
  }),
  playSoundWithState: jest.fn(),
  preload: jest.fn(),
  reset: jest.fn(),
  cleanup: jest.fn(),
  stop: jest.fn()
};

describe('<SpeakingTrainerContainer />', () => {
  const mockSpeechReset = jest.fn();

  const defaultProps = {
    locale: 'en_GB',
    learnLanguageAlpha3: 'RUS',
    learnLanguageBCP47: 'de-DE',
    displayLanguageBCP47: 'en-GB',
    mediaUrlService: {
      soundURL: jest.fn()
    },
    recordingService: {
      stop: jest.fn(),
      start: jest.fn(() => Promise.resolve())
    },
    speech: {
      permissionsGranted: true,
      ended: false,
      error: '',
      listening: false,
      start: jest.fn(),
      reset: mockSpeechReset,
      stop: jest.fn(),
      recording: false,
      transcript: 'test transcript'
    },
    micPermission: 'granted',
    completeItem: () => { },
    attemptItem: jest.fn(),
    onStart: jest.fn(),
    onFinish: jest.fn(),
    audio: mockAudio,
    trainer: {
      item_groups: [
        {
          items: [
            {
              id: 'a',
              sound: { id: 'aa' },
              learn_language_text: 'test a',
              display_language_text: 'TEST A'
            },
            {
              id: 'b',
              sound: { id: 'bb' },
              learn_language_text: 'test b',
              display_language_text: 'TEST B'
            },
            {
              id: 'c',
              sound: { id: 'cc' },
              learn_language_text: 'test c',
              display_language_text: 'TEST C'
            },
            {
              id: 'd',
              sound: { id: 'dd' },
              learn_language_text: 'test d (pl.)',
              display_language_text: 'TEST D'
            }
          ]
        }
      ]
    },
    track: jest.fn(),
    isMicOnboardingCompleted: false
  };

  describe('#mapStateToProps', () => {
    const state = {
      session: {
        learnLanguageAlpha3: 'RUS',
        locale: 'en_GB'
      },
      permissions: {
        micPermission: 'granted'
      },
      micOnboarding: {
        completed: true
      }
    };
    it('should match snapshot', () => {
      expect(mapStateToProps(state)).toEqual({
        micPermission: 'granted',
        learnLanguageBCP47: 'ru-RU',
        learnLanguageAlpha3: 'RUS',
        locale: 'en_GB'
      });
    });
  });

  describe('trainer item formatting', () => {
    let instance;

    beforeEach(() => {
      formatDL.mockClear();
      formatParentheses.mockClear();
      getFirstCorrectSolution.mockClear();
      removeGapFormatting.mockClear();
      markupStringToPlainText.mockClear();
      stripCurlyBracesAndContent.mockClear();

      const component = shallow(<SpeakingTrainerContainer {...defaultProps} />);
      instance = component.instance();
    });

    test('calls formatting methods', () => {
      expect(formatDL).toHaveBeenCalledTimes(instance.state.items.length);
      expect(formatParentheses).toHaveBeenCalledTimes(instance.state.items.length);
      expect(getFirstCorrectSolution).toHaveBeenCalledTimes(instance.state.items.length);
      expect(removeGapFormatting).toHaveBeenCalledTimes(instance.state.items.length);
      expect(markupStringToPlainText).toHaveBeenCalledTimes(instance.state.items.length);
      expect(stripCurlyBracesAndContent).toHaveBeenCalledTimes(instance.state.items.length);
    });

    test('strips parentheses and content', () => {
      expect(instance.state.items[3].learnLanguageText).toEqual('test d');
    });
  });

  describe('#handleAttempt', () => {
    it('should store attempt in state', () => {
      const component = shallow(<SpeakingTrainerContainer
        {...defaultProps}
      />);

      const attempt = {
        text: 'foo',
        solved: false,
        confidenceScore: 0.8501
      };

      component.instance().handleAttempt(attempt);
      expect(component.state('attempt')).toEqual({ mistakes: 2, text: 'foo' });

      component.instance().handleAttempt({ ...attempt, solved: true });
      expect(component.state('attempt')).toEqual({ mistakes: 0, text: 'foo' });
    });

    it('should play the feedback sound', () => {
      const component = shallow(<SpeakingTrainerContainer
        {...defaultProps}
      />);

      const attempt = {
        text: 'foo',
        solved: false,
        confidenceScore: 0.8501
      };

      component.instance().handleAttempt(attempt);
      expect(playFeedbackSound).toHaveBeenCalledTimes(1);
    });
  });

  describe('when speech recognition ends', () => {
    describe('when mic onboarding is completed', () => {
      it('should handle speech end', () => {
        const component = shallow(<SpeakingTrainerContainer
          {...defaultProps}
          isMicOnboardingCompleted={true}
        />);
        component.instance().setState({ micOnboardingComplete: true });
        const instance = component.instance();
        instance.handleSpeechEnd = jest.fn();

        component.setProps({ speech: { ...defaultProps.speech, ended: true } });
        expect(instance.handleSpeechEnd).toHaveBeenCalled();
      });
    });
  });

  describe('#handleSpeechEnd', () => {
    it('stops recordingService', () => {
      const component = shallow(<SpeakingTrainerContainer {...defaultProps} />);
      component.instance().handleSpeechEnd();
      expect(defaultProps.recordingService.stop).toHaveBeenCalled();
    });

    describe('speech ends with error or no transcript', () => {
      let instance;
      beforeEach(() => {
        const component = shallow(<SpeakingTrainerContainer
          {...defaultProps}
          speech={{
            ...defaultProps.speech,
            error: 'no-speech',
            transcript: ''
          }}
        />);
        instance = component.instance();
      });

      it('should not attempt item', () => {
        instance.handleSpeechEnd();
        expect(defaultProps.attemptItem).not.toHaveBeenCalled();
      });

      it('should not call handleAttempt method', () => {
        instance.handleAttempt = jest.fn();

        instance.handleSpeechEnd();
        expect(instance.handleAttempt).not.toHaveBeenCalled();
      });
    });

    describe('speech ends with transcript', () => {
      it('should attempt item', () => {
        const component = shallow(<SpeakingTrainerContainer
          {...defaultProps}
          speech={{
            ...defaultProps.speech,
            transcript: 'foo',
            solved: false,
            confidenceScore: 0.8501,
            engineName: 'WebAPISpeechRecognition'
          }}
        />);
        const item = {
          id: 'a',
          sound: { id: 'aa' },
          learnLanguageText: 'test a',
          displayLanguageText: 'TEST A'
        };
        const expectedAttempt = {
          text: 'foo',
          solved: false
        };

        component.instance().handleSpeechEnd();

        expect(defaultProps.attemptItem).toHaveBeenCalledWith(item, expectedAttempt);
      });

      it('should call handleAttempt method', () => {
        const component = shallow(<SpeakingTrainerContainer
          {...defaultProps}
          speech={{
            ...defaultProps.speech,
            transcript: 'testing 123',
            solved: true,
            confidenceScore: 0.8501,
            engineName: 'WebAPISpeechRecognition'
          }}
        />);

        const instance = component.instance();
        instance.handleAttempt = jest.fn(() => Promise.resolve());

        component.instance().handleSpeechEnd();
        expect(instance.handleAttempt).toHaveBeenCalledWith({ text: 'testing 123', solved: true });
      });
    });
  });

  describe('when speech error changes', () => {
    describe('when mic onboarding is completed', () => {
      let instance;

      beforeEach(() => {
        const component = shallow(<SpeakingTrainerContainer
          {...defaultProps}
          isMicOnboardingCompleted={true}
        />);

        instance = component.instance();
        instance.handleSpeechError = jest.fn();

        component.setProps({ speech: { ...defaultProps.speech, error: 'no-speech' } });
      });

      it('should handle speech error', () => {
        expect(instance.handleSpeechError).toHaveBeenCalled();
      });
    });
  });

  describe('#handleSpeechError', () => {
    it('attempts item with an error', () => {
      const component = shallow(<SpeakingTrainerContainer
        {...defaultProps}
        speech={{
          ...defaultProps.speech,
          engineName: 'WebAPISpeechRecognition',
          error: 'mock-error-language-not-supported'
        }}
      />);

      const item = {
        id: 'a',
        sound: { id: 'aa' },
        learnLanguageText: 'test a',
        displayLanguageText: 'TEST A'
      };
      const expectedAttempt = {
        engineName: 'WebAPISpeechRecognition',
        solved: false,
        error: 'mock-error-language-not-supported'
      };

      component.instance().handleSpeechError();
      expect(defaultProps.attemptItem).toHaveBeenCalledWith(item, expectedAttempt);
    });
  });

  describe('startSpeech', () => {
    let component;

    beforeEach(() => {
      defaultProps.speech.start.mockClear();
      defaultProps.audio.stop.mockClear();
    });

    describe('when not playing and not recording/listening', () => {
      beforeEach(() => {
        component = shallow(<SpeakingTrainerContainer {...defaultProps} />);
      });

      test('calls start speech', () => {
        component.instance().startSpeech();
        expect(defaultProps.speech.start).toHaveBeenCalled();
      });
    });

    describe('when audio is playing', () => {
      beforeEach(() => {
        component = shallow(<SpeakingTrainerContainer
          {...defaultProps}
          audio={{
            ...defaultProps.audio,
            isPlaying: true
          }}
        />);
      });

      test('should call audio stop', () => {
        component.instance().startSpeech();
        expect(defaultProps.audio.stop).toHaveBeenCalled();
      });
    });

    describe('when speech is listening or recording', () => {
      beforeEach(() => {
        component = shallow(<SpeakingTrainerContainer
          {...defaultProps}
          speech={{
            ...defaultProps.speech,
            recording: true,
            listening: false
          }}
        />);
      });

      test('should not start speech', () => {
        component.instance().startSpeech();
        expect(defaultProps.speech.start).not.toHaveBeenCalled();
      });
    });
  });

  describe('#handleAttemptRetry', () => {
    let component;
    let instance;

    beforeEach(() => {
      component = shallow(<SpeakingTrainerContainer {...defaultProps} />);
      instance = component.instance();
      instance.startSpeech = jest.fn();

      instance.handleAttemptRetry();
    });

    test('calls speech.reset', () => {
      expect(mockSpeechReset).toHaveBeenCalled();
    });

    test('calls startSpeech', () => {
      expect(instance.startSpeech).toHaveBeenCalled();
    });
  });

  describe('when onFeedbackAttemptCardClick callback is fired', () => {
    describe('when onboarding is not complete', () => {
      it('tracks gui:interacted', () => {
        const component = shallow(<SpeakingTrainerContainer {...defaultProps} />);
        component.instance().setState({ onboardingComplete: false, micOnboardingComplete: true });
        const onFeedbackAttemptCardClick = component.first().prop('onFeedbackAttemptCardClick');
        onFeedbackAttemptCardClick();
        expect(defaultProps.track).toHaveBeenCalledTimes(1);
        expect(defaultProps.track.mock.calls[0][0]).toMatchSnapshot();
      });
    });
  });

  describe('when onItemPlaybackCardClick callback is fired', () => {
    describe('when onboarding is not complete', () => {
      it('tracks gui:interacted', () => {
        const component = shallow(<SpeakingTrainerContainer {...defaultProps} />);
        component.instance().setState({ onboardingComplete: false, micOnboardingComplete: true });
        const onItemPlaybackCardClick = component.first().prop('onItemPlaybackCardClick');
        onItemPlaybackCardClick();
        expect(defaultProps.track).toHaveBeenCalledTimes(1);
        expect(defaultProps.track.mock.calls[0][0]).toMatchSnapshot();
      });
    });
  });

  describe('#handleMicButtonClick', () => {
    let component;
    let instance;

    beforeEach(() => {
      component = shallow(<SpeakingTrainerContainer {...defaultProps} />);
      instance = component.instance();
      instance.startSpeech = jest.fn();
    });

    afterEach(() => {
      defaultProps.track.mockClear();
    });

    describe('when onboarding is not complete', () => {
      beforeEach(() => {
        component.setProps({
          isMicOnboardingCompleted: false
        });
      });

      describe('when the first item is displayed', () => {
        beforeEach(() => {
          instance.setState({ currentIndex: 0 });
          instance.handleMicButtonClick();
        });

        it('starts speech', () => {
          expect(instance.startSpeech).toHaveBeenCalled();
        });

        it('tracks gui:interacted', () => {
          expect(defaultProps.track).toHaveBeenCalledTimes(1);
          expect(defaultProps.track.mock.calls[0][0]).toMatchSnapshot();
        });
      });

      describe('when not the first item is displayed', () => {
        beforeEach(() => {
          instance.setState({ currentIndex: 1 });
          instance.handleMicButtonClick();
        });

        it('starts speech', () => {
          expect(instance.startSpeech).toHaveBeenCalled();
        });

        it('does not send tracking event', () => {
          expect(defaultProps.track).not.toHaveBeenCalled();
        });
      });
    });

    describe('when onboarding is complete', () => {
      beforeEach(() => {
        instance.setState({ currentIndex: 0 });
        component.setProps({
          isMicOnboardingCompleted: true
        });

        instance.setState({ onboardingCompleted: true });
        instance.handleMicButtonClick();
      });

      it('starts speech', () => {
        expect(instance.startSpeech).toHaveBeenCalled();
      });

      it('does not send tracking event', () => {
        expect(defaultProps.track).not.toHaveBeenCalled();
      });
    });
  });
});
