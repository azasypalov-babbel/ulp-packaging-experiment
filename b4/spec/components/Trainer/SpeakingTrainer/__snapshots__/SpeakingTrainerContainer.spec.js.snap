// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpeakingTrainerContainer /> #handleMicButtonClick when onboarding is not complete when the first item is displayed tracks gui:interacted 1`] = `
Object {
  "event": "gui:interacted",
  "payload": Object {
    "gui_container": "speaking_trainer",
    "gui_element": "mic_button",
    "interaction": "clicked",
    "learn_language_alpha3": "RUS",
    "locale": "en_GB",
    "origin": "lesson_player",
  },
  "version": 1,
}
`;

exports[`<SpeakingTrainerContainer /> when onFeedbackAttemptCardClick callback is fired when onboarding is not complete tracks gui:interacted 1`] = `
Object {
  "event": "gui:interacted",
  "payload": Object {
    "gui_container": "speaking_trainer",
    "gui_element": "feedback_attempt_card",
    "interaction": "clicked",
    "learn_language_alpha3": "RUS",
    "locale": "en_GB",
    "origin": "lesson_player",
  },
  "version": 1,
}
`;

exports[`<SpeakingTrainerContainer /> when onItemPlaybackCardClick callback is fired when onboarding is not complete tracks gui:interacted 1`] = `
Object {
  "event": "gui:interacted",
  "payload": Object {
    "gui_container": "speaking_trainer",
    "gui_element": "item_playback_card",
    "interaction": "clicked",
    "learn_language_alpha3": "RUS",
    "locale": "en_GB",
    "origin": "lesson_player",
  },
  "version": 1,
}
`;
